### PR TITLE
chore: Add annotation to restrict API usage

### PR DESCRIPTION
**Pull request by eeatonaws**
- [x] PR title and description conform to [Pull Request](https://github.com/aws-amplify/amplify-android/blob/main/CONTRIBUTING.md#pull-request-guidelines) guidelines.

*Issue #, if available:* N/A

*Description of changes:* Add an annotation similar to Android's RestrictTo since RestrictTo does not work for Kotlin classes and APIs.

*How did you test these changes?*
Tested the annotation in a sample app.

*Documentation update required?*
- [x] No
- [ ] Yes (Please include a PR link for the documentation update)

*General Checklist*
- [ ] Added Unit Tests
- [ ] Added Integration Tests
- [x] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by div5yesh — DISMISSED** [review]

File: annotations/src/main/java/com/amplifyframework/annotations/RestrictToAmplify.kt [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+)
+@Target(
+    AnnotationTarget.CLASS,
+    AnnotationTarget.TYPEALIAS,
+    AnnotationTarget.FUNCTION,
+    AnnotationTarget.PROPERTY,
+    AnnotationTarget.FIELD,
+    AnnotationTarget.CONSTRUCTOR,
+)
+public annotation class RestrictToAmplify
```
Comment (mattcreaser) [inline annotations/src/main/java/com/amplifyframework/annotations/RestrictToAmplify.kt:37]: Can we go with `InternalAmplifyApi` - that would be more consistent with other libraries.

Examples:
[InternalCoroutinesApi](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-internal-coroutines-api/)
[InternalComposeApi](https://developer.android.com/reference/kotlin/androidx/compose/runtime/InternalComposeApi)
[InternalSdkApi](https://github.com/awslabs/aws-sdk-kotlin/blob/177b444b561fa99a6eda0b641307e462cb192cb8/aws-runtime/aws-core/common/src/aws/sdk/kotlin/runtime/Annotations.kt) - aws kotlin sdk

**Review by mattcreaser — CHANGES_REQUESTED** [review]

**Comment by mattcreaser** [comment]
Can we also change existing instances of RestrictTo in this PR?

**Comment by tylerjroach** [comment]
> Can we also change existing instances of RestrictTo in this PR?

@mattcreaser I had suggested leaving RestrictTo in for for Java. Thoughts?

**Comment by eeatonaws** [comment]
> > Can we also change existing instances of RestrictTo in this PR?
> 
> @mattcreaser I had suggested leaving RestrictTo in for for Java. Thoughts?

Instances of RestrictTo have been replaced with InternalAmplifyApi.

**Review by tylerjroach — APPROVED** [review]

**Review by mattcreaser — APPROVED** [review]

**Review by sdhuka — APPROVED** [review]

**Pull request merged**